src/test/spec/trace.rs

```diff
@@ -0,0 +1,215 @@
+        tracing::Level::DEBUG,
+    )]));
+    let mut tracing_subscriber = DEFAULT_GLOBAL_TRACING_HANDLER.subscribe();
+
+    let docs = iter::repeat(doc! { "x": "y" }).take(100);
+    coll.insert_many(docs, None)
+        .await
+        .expect("insert many should succeed");
+
+    let events = tracing_subscriber
+        .collect_events(Duration::from_millis(500), |_| true)
+        .await;
+    assert_eq!(events.len(), 2);
+
+    let started = &events[0];
+    let command = started.get_value_as_string("command");
+    assert_eq!(command.len(), DEFAULT_MAX_DOCUMENT_LENGTH_BYTES + 3); // +3 for trailing "..."
+
+    let succeeded = &events[1];
+    let reply = succeeded.get_value_as_string("reply");
+    assert!(reply.len() <= DEFAULT_MAX_DOCUMENT_LENGTH_BYTES + 3); // +3 for trailing "..."
+
+    coll.find(None, None).await.expect("find should succeed");
+    let succeeded = tracing_subscriber
+        .wait_for_event(Duration::from_millis(500), |e| {
+            e.get_value_as_string("message") == "Command succeeded"
+        })
+        .await
+        .unwrap();
+    let reply = succeeded.get_value_as_string("reply");
+    assert_eq!(reply.len(), DEFAULT_MAX_DOCUMENT_LENGTH_BYTES + 3); // +3 for trailing "..."
+}
+
+/// Prose test 2: explicitly configured truncation limit
+#[cfg_attr(feature = "tokio-runtime", tokio::test)]
+#[cfg_attr(feature = "async-std-runtime", async_std::test)]
+async fn command_logging_truncation_explicit_limit() {
+    let _guard = LOCK.run_exclusively().await;
+
+    let mut client_opts = CLIENT_OPTIONS.get().await.clone();
+    client_opts.tracing_max_document_length_bytes = Some(5);
+    let client = TestClient::with_options(Some(client_opts)).await;
+
+    let _levels_guard = DEFAULT_GLOBAL_TRACING_HANDLER.set_levels(HashMap::from([(
+        COMMAND_TRACING_EVENT_TARGET.to_string(),
+        tracing::Level::DEBUG,
+    )]));
+    let mut tracing_subscriber = DEFAULT_GLOBAL_TRACING_HANDLER.subscribe();
+
+    client
+        .database("tracing_test")
+        .run_command(doc! { "hello" : "true" }, None)
+        .await
+        .expect("hello command should succeed");
+
+    let events = tracing_subscriber
+        .collect_events(Duration::from_millis(500), |_| true)
+        .await;
+    assert_eq!(events.len(), 2);
+
+    let started = &events[0];
+    let command = started.get_value_as_string("command");
+    assert_eq!(command.len(), 8); // 5 + 3 for trailing "..."
+
+    let succeeded = &events[1];
+    let reply = succeeded.get_value_as_string("reply");
+    assert_eq!(reply.len(), 8); // 5 + 3 for trailing "..."
+
+    // TODO RUST-1405: when we expose the full server reply for command errors, we should confirm
+    // that gets correctly truncated in command failed events here as well.
+}
+
+/// Prose test 3: mid-codepoint truncation
+#[cfg_attr(feature = "tokio-runtime", tokio::test)]
+#[cfg_attr(feature = "async-std-runtime", async_std::test)]
+async fn command_logging_truncation_mid_codepoint() {
```

Comment:
this test is unfortunately really brittle to changes in the server response, but I don't think there's really a way around that